backend/recommendation.js

```diff
@@ -0,0 +1,91 @@
+const { PrismaClient } = require('@prisma/client');
+const prisma = new PrismaClient();
+
+function generateTimeGrid() {
+    const timeGrid = [];
+    for (let i = 0; i < 48; i++) {
+        timeGrid.push({ slot: i, available: true });
+    }
+    return timeGrid;
+}
+
+function markUnavailableSlots(timeGrid, startAt, endAt) {
+    const startSlot = Math.floor(startAt.getHours() * 2 + startAt.getMinutes() / 30);
+    const endSlot = Math.ceil(endAt.getHours() * 2 + endAt.getMinutes() / 30);
+    for (let i = startSlot; i < endSlot; i++) {
+        timeGrid[i].available = false;
+    }
+}
+
+function findAvailableSlots(timeGrid, duration) {
+    const availableSlots = [];
+    let consecutiveSlots = 0;
+    for (let i = 0; i < timeGrid.length; i++) {
+        if (timeGrid[i].available) {
+            consecutiveSlots++;
+            if (consecutiveSlots >= duration) {
+                availableSlots.push(i - duration + 1);
+            }
+        } else {
+            consecutiveSlots = 0;
+        }
+    }
+    return availableSlots;
+}
+
+async function recommendEventSlots(userId, duration, targetDate, invitees = []) {
+    try {
+        const parsedDate = new Date(targetDate);
+        const startOfDay = new Date(parsedDate.setHours(0, 0, 0, 0));
+        const endOfDay = new Date(parsedDate.setHours(23, 59, 59, 999));
+
+        const users = await prisma.user.findMany({
+            where: {
+                OR: [{ id: userId }, { email: { in: invitees } }]
+            },
+            include: {
+                calendarEvents: {
+                    where: {
+                        status: 'accepted',
+                        AND: [
+                            { startAt: { lte: endOfDay } },
```

Comment:
sorry, this might be a noob question but why is endOfDay set here for startAt? Same with the line below in reverse

Reply:
Apologies for not leaving descriptive enough comments here, what I am doing here is trying to get the calendar events which have a startAt that is less than endOfDay (which means that start before 11:59PM of the specified day) and also have an endAt that is greater than startOfDay (which means that end after 12:00AM of the specified day). When these two conditions are combined, it gives a list of calendar events that are essentially within the range of the specified day.

Hopefully that clarifies what I did